compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs

```diff
@@ -557,12 +558,114 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                         _ => "",
                     };
 
-                    multi_suggestions.push(SourceKindMultiSuggestion::new_fully_qualified(
-                        receiver.span,
-                        def_path,
-                        adjustment,
-                        successor,
-                    ));
+                    let mut paths = vec![];
+                    let param_env = ty::ParamEnv::reveal_all();
+                    let name = self.infcx.tcx.item_name(def_id);
+                    // Look for all the possible implementations to suggest, otherwise we'll show
+                    // just suggest the syntax for the fully qualified path with placeholders.
+                    self.infcx.tcx.for_each_relevant_impl(
```

Comment:
This logic is kinda large and deeply nested. Please pull it into a helper method.